From d2678c4f0ae509f1587284cd80c8ee5a8db6daea Mon Sep 17 00:00:00 2001
From: Peter Chen <peter.chen@nxp.com>
Date: Tue, 19 Jun 2018 10:17:27 +0800
Subject: [PATCH 4142/5242] MLK-18609-4 usb: cdns3: gadget: refine de-init
 process

commit  20e7a4687529c6c7ee10af60db145a8656c014d3 from
https://source.codeaurora.org/external/imx/linux-imx.git

- The gadget speed should be reset to USB_SPEED_UNKNOWN at any de-init
  process
- The TRB buffer should be free when the gadget is removed

Reviewed-by: Jun Li <jun.li@nxp.com>
Signed-off-by: Peter Chen <peter.chen@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/usb/cdns3/gadget.c |    8 +++++---
 1 file changed, 5 insertions(+), 3 deletions(-)

diff --git a/drivers/usb/cdns3/gadget.c b/drivers/usb/cdns3/gadget.c
index 3b007ce..9e3e011 100644
--- a/drivers/usb/cdns3/gadget.c
+++ b/drivers/usb/cdns3/gadget.c
@@ -2069,6 +2069,10 @@ static int usb_ss_gadget_udc_stop(struct usb_gadget *gadget)
 
 	usb_ss->onchip_mem_allocated_size = 0;
 	usb_ss->out_mem_is_allocated = 0;
+	usb_ss->gadget.speed = USB_SPEED_UNKNOWN;
+	for (i = 0; i < usb_ss->ep_nums ; i++)
+		usb_ss_free_trb_pool(usb_ss->eps[i]);
+
 	if (!usb_ss->start_gadget)
 		return 0;
 
@@ -2086,9 +2090,6 @@ static int usb_ss_gadget_udc_stop(struct usb_gadget *gadget)
 	gadget_writel(usb_ss, &usb_ss->regs->usb_ien, 0);
 	gadget_writel(usb_ss, &usb_ss->regs->usb_conf, USB_CONF__DEVDS__MASK);
 
-	for (i = 0; i < usb_ss->ep_nums ; i++)
-		usb_ss_free_trb_pool(usb_ss->eps[i]);
-
 	return ret;
 }
 
@@ -2418,6 +2419,7 @@ static void __cdns3_gadget_stop(struct cdns3 *cdns)
 		usb_ss->gadget_driver->disconnect(&usb_ss->gadget);
 	usb_gadget_disconnect(&usb_ss->gadget);
 	spin_lock_irqsave(&usb_ss->lock, flags);
+	usb_ss->gadget.speed = USB_SPEED_UNKNOWN;
 	/* disable interrupt for device */
 	gadget_writel(usb_ss, &usb_ss->regs->usb_ien, 0);
 	gadget_writel(usb_ss, &usb_ss->regs->usb_conf, USB_CONF__DEVDS__MASK);
-- 
1.7.9.5

